src/rusty/src/dns_headers.rs

```diff
@@ -0,0 +1,128 @@
+use std::net::{IpAddr, Ipv6Addr, ToSocketAddrs};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+use std::panic::catch_unwind;
+
+use std::ffi::CStr;
+use std::os::raw::c_char;
+
+use crate::bridge::*;
+use crate::await_ibd_complete_or_stalled;
+
+/// Maps a set of six IPv6 addresses to an 80-byte Bitcoin header.
+/// The first two bytes of each address are ignored.
+/// The next 4 bits in each address indicate the ordering of the addresses
+/// (as DNS resolvers/servers often shuffle the addresses)
+/// The first 8 bits (ie the second half of the 3rd byte and first half of the 4th)
+/// of the first address are interpreted as a version and must currently be 0.
+/// The remaining bits are placed into the 80 byte result in order.
+fn map_addrs_to_header(ips: &mut [Ipv6Addr]) -> [u8; 80] {
+    let mut header = [0u8; 80];
+    if ips.len() != 6 { return header; }
+    ips.sort_unstable_by(|a, b| {
+        // Sort based on the first 4 bits in the 3rd byte...
+        (&(a.octets()[2] & 0xf0)).cmp(&(b.octets()[2] & 0xf0))
+    });
+    if ips.len() != 6 { unreachable!(); }
+    let version = (ips[0].octets()[2] & 0x0f) | (ips[0].octets()[3] & 0xf0);
+    if version != 0 { return header; }
+
+    let mut offs = 0; // in bytes * 2
+    for (idx, ip) in ips.iter().enumerate() {
+        for i in if idx == 0 { 3..14*2 } else { 1..14*2 } {
+            if i % 2 == 1 {
+                header[offs/2] |= (ip.octets()[i/2 + 2] & 0x0f) >> 0;
+            } else {
+                header[offs/2] |= (ip.octets()[i/2 + 2] & 0xf0) >> 4;
+            }
+            if offs % 2 == 0 {
+                header[offs/2] <<= 4;
+            }
+            offs += 1;
+        }
+    }
+    header
+}
+
+#[test]
```

Comment:
feature gate this? `#[cfg(test)]`

Reply:
Done. Double checked that you can slip non-compiling code into the function and it won't fail to build until you run make check, and that make check still fails if you change the header.